Layout

nicoddemus opened the pull request:
Moved tests outside of the `qutebrowser` package, in doing helpers and stubs modules were no longer importable, so I took that opportunity to change helpers and stubs functionality into fixtures and convert the tests that were using those modules to `pytest`.
- Introduced [pytest-mock](https://github.com/pytest-dev/pytest-mock) plugin;
- In order to do not pollute the PR with more changes than necessary, I kept tests in classes and converted `setUp/tearDown` pairs into autouse fixtures defined in the test classes themselves.

nicoddemus committed:
Implemented test for TextBase widget

nicoddemus committed:
Implemented test for Percentage widget

nicoddemus committed:
Implemented test for Progress widget

Also created a conftest file with a "default_config" fixture.

nicoddemus committed:
Fixing docstring typo in test_progress

nicoddemus committed:
Moving logging and QApplication to conftest

As discussed in #8

nicoddemus committed:
Converted test_standarddir to pytest

 related to #10

nicoddemus committed:
Moved tests to outside of qutebrowser package

nicoddemus committed:
Created stubs fixture and converted test_stubs to pytest

nicoddemus committed:
Removed environ_set_temp as we will use monkeypatch

nicoddemus committed:
Converted test_webelem to pytest

nicoddemus committed:
Converted test_content_disposition to pytest

nicoddemus committed:
Converted test_http to pytest

nicoddemus committed:
Adding pytest-mock to tox

nicoddemus committed:
Converted test_conftest to pytest

nicoddemus committed:
Converted test_tabhistory to pytest

nicoddemus committed:
Converted test_config to pytest

nicoddemus committed:
Removing (object) subclassing from test classes

Old habit from python 2

nicoddemus committed:
Converted test_basekeyparser to pytest

nicoddemus committed:
Converted test_readline to pytest

nicoddemus committed:
Renamed singleShot to isSingleShot in QTimer stub

As in compliance to: http://doc.qt.io/qt-5/qtimer.html

nicoddemus committed:
Converted test_modeparsers to pytest

nicoddemus committed:
Converted test_editor to pytest

nicoddemus committed:
Converted test_debug to pytest

nicoddemus committed:
Converted test_urlutils to pytest

nicoddemus committed:
Converted test_utils to pytest

nicoddemus committed:
Removed unnecessary module 'helpers'

nicoddemus commented on test/config/test_config.py:
```diff
@@ -116,93 +114,88 @@ def test_interpolation_cross_section(self):
     """Test validating of the default config."""
 
+    @pytest.mark.xfail(reason='fails if run with other tests in this module, '
+                              'succeeds if executed alone (The-Compiler, help)'
```
@The-Compiler I would like some advice here, please. :smile:

The-Compiler replied:
Congrats, you spotted a bug! :laughing: 

`config.ConfigManager` initializes its default values from `configdata.DATA`. That's a dict `{str: Section}`, and a section contains the values.

Now since `DATA` is a global, when multiple `ConfigManager`s are used, they actually _share_ their sections, which means the change from `test_invalid_interpolation` gets carried over, tested and fails.

I'll need to check what the best fix for this is - probably using a factory function in `configdata` instead of a global `DATA`.

nicoddemus committed:
Merge remote-tracking branch 'origin/master' into layout

Conflicts:
	qutebrowser/test/keyinput/test_basekeyparser.py
	qutebrowser/test/utils/test_standarddir.py
	test/browser/http/test_content_disposition.py
	test/config/test_configtypes.py
	test/misc/test_editor.py
	test/utils/test_debug.py
	test/utils/test_utils.py
	tox.ini

The-Compiler commented:
I haven't looked into it yet in detail, but seems like there are some other failures as well:

```
___________________________________________________________________________________________ TestGetStandardDirLinux.test_data_explicit ____________________________________________________________________________________________

self = <test_standarddir.TestGetStandardDirLinux object at 0x7f7e8418c208>, monkeypatch = <_pytest.monkeypatch.monkeypatch object at 0x7f7e8418cd68>, tmpdir = local('/tmp/pytest-1/test_data_explicit0')

    def test_data_explicit(self, monkeypatch, tmpdir):
        """Test data dir with XDG_DATA_HOME explicitely set."""
        monkeypatch.setenv('XDG_DATA_HOME', str(tmpdir))
        standarddir.init(None)
>       assert standarddir.data() == str(tmpdir / 'qutebrowser')
E       assert '/tmp/pytest-...ebrowser_test' == '/tmp/pytest-1...0/qutebrowser'
E         Skipping 34 identical leading characters in diff, use -v to show
E         - qutebrowser_test
E         ?            -----
E         + qutebrowser

test/utils/test_standarddir.py:54: AssertionError
__________________________________________________________________________________________ TestGetStandardDirLinux.test_config_explicit ___________________________________________________________________________________________

self = <test_standarddir.TestGetStandardDirLinux object at 0x7f7ed0971c50>, monkeypatch = <_pytest.monkeypatch.monkeypatch object at 0x7f7ee2990b70>, tmpdir = local('/tmp/pytest-1/test_config_explicit0')

    def test_config_explicit(self, monkeypatch, tmpdir):
        """Test config dir with XDG_CONFIG_HOME explicitely set."""
        monkeypatch.setenv('XDG_CONFIG_HOME', str(tmpdir))
        standarddir.init(None)
>       assert standarddir.config() == str(tmpdir / 'qutebrowser')
E       assert '/tmp/pytest-...ebrowser_test' == '/tmp/pytest-1...0/qutebrowser'
E         Skipping 36 identical leading characters in diff, use -v to show
E         - qutebrowser_test
E         ?            -----
E         + qutebrowser

test/utils/test_standarddir.py:60: AssertionError
___________________________________________________________________________________________ TestGetStandardDirLinux.test_cache_explicit ___________________________________________________________________________________________

self = <test_standarddir.TestGetStandardDirLinux object at 0x7f7e84589780>, monkeypatch = <_pytest.monkeypatch.monkeypatch object at 0x7f7e845895f8>, tmpdir = local('/tmp/pytest-1/test_cache_explicit0')

    def test_cache_explicit(self, monkeypatch, tmpdir):
        """Test cache dir with XDG_CACHE_HOME explicitely set."""
        monkeypatch.setenv('XDG_CACHE_HOME', str(tmpdir))
        standarddir.init(None)
>       assert standarddir.cache() == str(tmpdir / 'qutebrowser')
E       assert '/tmp/pytest-...ebrowser_test' == '/tmp/pytest-1...0/qutebrowser'
E         Skipping 35 identical leading characters in diff, use -v to show
E         - qutebrowser_test
E         ?            -----
E         + qutebrowser

test/utils/test_standarddir.py:66: AssertionError
________________________________________________________________________________________________ TestGetStandardDirLinux.test_data ________________________________________________________________________________________________

self = <test_standarddir.TestGetStandardDirLinux object at 0x7f7e8458f2e8>, monkeypatch = <_pytest.monkeypatch.monkeypatch object at 0x7f7e8458f358>, tmpdir = local('/tmp/pytest-1/test_data0')

    def test_data(self, monkeypatch, tmpdir):
        """Test data dir with XDG_DATA_HOME not set."""
        monkeypatch.setenv('HOME', str(tmpdir))
        monkeypatch.setenv('XDG_DATA_HOME', None)
        standarddir.init(None)
        expected = tmpdir / '.local' / 'share' / 'qutebrowser'
>       assert standarddir.data() == str(expected)
E       assert 'None/qutebrowser_test' == '/tmp/pytest-1/te...share/qutebrowser'
E         - None/qutebrowser_test
E         + /tmp/pytest-1/test_data0/.local/share/qutebrowser

test/utils/test_standarddir.py:74: AssertionError
_______________________________________________________________________________________________ TestGetStandardDirLinux.test_config _______________________________________________________________________________________________

self = <test_standarddir.TestGetStandardDirLinux object at 0x7f7e842ebfd0>, monkeypatch = <_pytest.monkeypatch.monkeypatch object at 0x7f7e842eb518>, tmpdir = local('/tmp/pytest-1/test_config0')

    def test_config(self, monkeypatch, tmpdir):
        """Test config dir with XDG_CONFIG_HOME not set."""
        monkeypatch.setenv('HOME', str(tmpdir))
        monkeypatch.setenv('XDG_CONFIG_HOME', None)
        standarddir.init(None)
        expected = tmpdir / '.config' / 'qutebrowser'
>       assert standarddir.config() == str(expected)
E       assert 'None/qutebrowser_test' == '/tmp/pytest-1/te...onfig/qutebrowser'
E         - None/qutebrowser_test
E         + /tmp/pytest-1/test_config0/.config/qutebrowser

test/utils/test_standarddir.py:82: AssertionError
_______________________________________________________________________________________________ TestGetStandardDirLinux.test_cache ________________________________________________________________________________________________

self = <test_standarddir.TestGetStandardDirLinux object at 0x7f7e8457da20>, monkeypatch = <_pytest.monkeypatch.monkeypatch object at 0x7f7e8457de10>, tmpdir = local('/tmp/pytest-1/test_cache0')

    def test_cache(self, monkeypatch, tmpdir):
        """Test cache dir with XDG_CACHE_HOME not set."""
        monkeypatch.setenv('HOME', str(tmpdir))
        monkeypatch.setenv('XDG_CACHE_HOME', None)
        standarddir.init(None)
        expected = tmpdir / '.cache' / 'qutebrowser'
>       assert standarddir.cache() == expected
E       assert 'None/qutebrowser_test' == local('/tmp/pytest-1/test_cache0/.cache/qutebrowser')
E        +  where 'None/qutebrowser_test' = <function cache at 0x7f7ed1dfad90>()
E        +    where <function cache at 0x7f7ed1dfad90> = standarddir.cache

test/utils/test_standarddir.py:90: AssertionError
___________________________________________________________________________________________________ TestReadFile.test_readfile ____________________________________________________________________________________________________

self = <test_utils.TestReadFile object at 0x7f7ee294f6a0>

    def test_readfile(self):
        """Read a testfile."""
        directory = os.path.dirname(__file__)
>       content = utils.read_file(os.path.join(directory, 'testfile'))

test/utils/test_utils.py:75: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
qutebrowser/utils/utils.py:89: in read_file
    data = pkg_resources.resource_string(qutebrowser.__name__, filename)
.tox/unittests/lib/python3.4/site-packages/pkg_resources/__init__.py:1139: in resource_string
    self, resource_name
.tox/unittests/lib/python3.4/site-packages/pkg_resources/__init__.py:1567: in get_resource_string
    return self._get(self._fn(self.module_path, resource_name))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <pkg_resources.DefaultProvider object at 0x7f7e8418c748>, path = '/home/florian/proj/qutebrowser/hackebrot-git/qutebrowser/home/florian/proj/qutebrowser/hackebrot-git/test/utils/testfile'

    def _get(self, path):
>       with open(path, 'rb') as stream:
            return stream.read()
E           FileNotFoundError: [Errno 2] No such file or directory: '/home/florian/proj/qutebrowser/hackebrot-git/qutebrowser/home/florian/proj/qutebrowser/hackebrot-git/test/utils/testfile'

.tox/unittests/lib/python3.4/site-packages/pkg_resources/__init__.py:1690: FileNotFoundError
```

nicoddemus commented:
About the failing tests in `test_standarddir`, I'm not sure why they're failing but I didn't run them as I use Windows and don't have a Linux box handy at the moment, so if you or @hackebrot could help find out what the problem is I would be grateful. :smile:

As for `test_utils.TestReadFile`, I just realized it tries to load resources from the `qutebrowser` package... where should we put `testfile` so `read_file` can find it them? At the root of the package, perhaps?

The-Compiler commented:
I fixed the config test in 57158e719135c97fd4d09ba38553cb3eac589649 and updated your master.

I also rebased this branch on master to verify the fix works... then I spent way too much time resolving merge conflicts, mostly because of the spelling fixes in 8ebac8d38c32bbc028661130fc1ffe746fae9cc4 - I wasn't realizing you already started working on this before them. I'm glad I have [git rerere](http://git-scm.com/book/en/v2/Git-Tools-Rerere) turned on, so at least I don't have to resolve them again. There were also some conflicts where I didn't exactly know what was going on though.... Well, I hope the second attempt will work better.

For `TestReadFile`, I'd say let's put the file into `utils/` - less visible and it gets used from `utils.py`.

I'm looking into the `test_standarddir` failures now.

The-Compiler committed:
Fix paths for TestGetStandardDirLinux.

The-Compiler committed:
Fix deleting of envvars in TestGetStandardDirLinux

nicoddemus committed:
Moving testfile used by test_utils into qutebrowser.utils

nicoddemus commented:
I moved `tesfile` to `utils/` like you suggested. :)

Let me know if you have any questions regarding my changes in `test_standarddir`

The-Compiler commented:
Oh, I didn't notice you did a merge already - thanks!

For the standarddir tests on Linux, there were two issues:
- Before the changes, the tests on Windows used `qutebrowser_test` as Qt applicationName as we can't easily set `XDG_*_HOME` to a tempdir there - but the Linux tests used `qutebrowser` as it's in a temp dir anyways. After your change, both use the same fixture with `qutebrowser_test`, so I updated the assertions to match.
- My setenv helper understood `None` as a special value for "delete this environment variable" - `monkeypatch.setenv` instead set it to `str(None)` which caused that issue and created a `None` directory :laughing: - I adjusted them to use `monkeypatch.delenv` instead.

All tests now pass after your change :+1: - I'll now look at the diff and either ask some more questions or merge :smile:

If I understand it correctly, the changes from #9 are contained in here as well, right?

nicoddemus commented:
> All tests now pass after your change

Excellent! :smile:

>  I'll now look at the diff and either ask some more questions or merge

Great! I understand it is a pretty big diff, even though I tried to keep the changes to a bare minimum, hehehe!

Cheers :beers:

The-Compiler commented on test/browser/test_webelem.py:
```diff
@@ -360,189 +360,190 @@ def test_multiple_levels(self):
+
+    @pytest.yield_fixture
+    def stub_config(self, stubs):
+        config = stubs.ConfigStub({'input': {'insert-mode-on-plugins': True}})
```
Why is `insert-mode-on-plugins` set here already, and then again later in the tests? Shouldn't this rather be something like `config = stubs.ConfigStub({'input': {}})` instead?

nicoddemus replied:
Agreed, fixed.

The-Compiler commented on test/config/test_config.py:
```diff
@@ -116,93 +114,88 @@ def test_interpolation_cross_section(self):
+    def setup(self, tmpdir):
+        self.conf_path = (tmpdir / 'config').ensure(dir=1)
+        self.data_path = (tmpdir / 'data').ensure(dir=1)
+        self.cache_path = (tmpdir / 'cache').ensure(dir=1)
```
What does the `.ensure(dir=1)` here do?

nicoddemus replied:
It will ensure the directory exists, creating it if necessary. [Reference](http://pylib.readthedocs.org/en/latest/path.html#py._path.local.LocalPath.ensure).

The-Compiler commented on test/config/test_configtypes.py:
```diff
@@ -70,72 +67,77 @@ def __repr__(self):
         """Test __iter__ without a description."""
         vv = configtypes.ValidValues('foo', 'bar')
-        self.assertEqual(list(vv), ['foo', 'bar'])
+        assert list(vv), ['foo' == 'bar']
```
I think this should be `assert list(vv) == ['foo', 'bar']` - if I parse this right, you're basically asserting `(list(vv), [False])` now, which is always true :stuck_out_tongue:

nicoddemus replied:
Argh, regex-replacement going wrong, hehehe. Fixed all.

The-Compiler commented on test/config/test_configtypes.py:
```diff
@@ -70,72 +67,77 @@ def __repr__(self):
         """Test __iter__ with a description."""
         vv = configtypes.ValidValues(('foo', "foo desc"), ('bar', "bar desc"))
-        self.assertEqual(list(vv), ['foo', 'bar'])
+        assert list(vv), ['foo' == 'bar']
```
Same as above.

The-Compiler commented on test/config/test_configtypes.py:
```diff
@@ -70,72 +67,77 @@ def __repr__(self):
         """Test __iter__ with mixed description."""
         vv = configtypes.ValidValues(('foo', "foo desc"), 'bar')
-        self.assertEqual(list(vv), ['foo', 'bar'])
+        assert list(vv), ['foo' == 'bar']
```
Same as above.

The-Compiler commented on test/config/test_configtypes.py:
```diff
@@ -147,40 +149,37 @@ def test_validate(self):
         """Test complete with valid_values set without description."""
         self.t.valid_values = configtypes.ValidValues('foo', 'bar')
-        self.assertEqual(self.t.complete(), [('foo', ''), ('bar', '')])
+        assert self.t.complete(), [('foo', ''), ('bar' == '')]
```
Same as above.

The-Compiler commented on test/config/test_configtypes.py:
```diff
@@ -469,30 +466,28 @@ def test_validate_empty_none_ok(self):
     def test_transform_empty(self):
         """Test transform with an empty value."""
-        self.assertEqual(self.t.transform('23,,42'), [23, None, 42])
+        assert self.t.transform('23,,42'), [23, None == 42]
```
Same as above.

The-Compiler commented on test/config/test_configtypes.py:
```diff
@@ -724,37 +719,37 @@ def test_validate_empty_none_ok(self):
         """Test transform with an empty value."""
-        self.assertEqual(self.t.transform('23%,,42%'), [23, None, 42])
-
+        assert self.t.transform('23%,,42%'), [23, None == 42]
```
Same as above.

The-Compiler commented on test/config/test_configtypes.py:
```diff
@@ -724,37 +719,37 @@ def test_validate_empty_none_ok(self):
     def test_transform_more(self):
         """Test transform with multiple values."""
-        self.assertEqual(self.t.transform('23%,42%,1337%'), [23, 42, 1337])
+        assert self.t.transform('23%,42%,1337%'), [23, 42 == 1337]
```
Same as above.

The-Compiler commented on test/config/test_configtypes.py:
```diff
@@ -1231,34 +1217,32 @@ def test_validate_empty_none_ok(self):
-            with self.subTest(val=val):
-                with self.assertRaises(configexc.ValidationError, msg=val):
-                    self.t.validate(val)
+        with pytest.raises(configexc.ValidationError, msg=val):
```
I think the `msg=val` here does nothing (and is unneeded, as pytest prints the parameter anyways I think)

The-Compiler commented on test/config/test_configtypes.py:
```diff
@@ -1600,32 +1582,32 @@ def test_validate_len_ok(self):
     def test_transform_more(self):
         """Test transform with multiple values."""
-        self.assertEqual(self.t.transform('23,2k,1337'), [23, 2048, 1337])
+        assert self.t.transform('23,2k,1337'), [23, 2048 == 1337]
```
And another one - sorry for the spam, but I'm going to comment every one I see so none gets missed :)

nicoddemus replied:
No please... that's just a regex/replace gone wrong, your detailed review is certainly appreciated. I will see to fix those tomorrow. :smile:

The-Compiler commented on test/config/test_configtypes.py:
```diff
@@ -1600,32 +1582,32 @@ def test_validate_len_ok(self):
         """Test transform with an empty value."""
-        self.assertEqual(self.t.transform('23,,42'), [23, None, 42])
-
+        assert self.t.transform('23,,42'), [23, None == 42]
```
Same.

The-Compiler commented on test/config/test_configtypes.py:
```diff
@@ -1645,37 +1627,37 @@ def test_validate_placeholder(self):
     def test_transform_double(self):
         """Test transform with two words."""
-        self.assertEqual(self.t.transform('foobar baz'), ['foobar', 'baz'])
+        assert self.t.transform('foobar baz'), ['foobar' == 'baz']
```
Same

The-Compiler committed:
Fix testfile path in MANIFEST.in.

The-Compiler commented:
The `test` folder should probably be named `tests` instead, to not mask the `test` package in python's stdlib.

I also started adjusting `tox.ini` so the linter tools run with the new location as well and noticed a problem: `pylint` seems to want a proper Python package with `__init__.py` files if you pass a package :disappointed: - a possible fix would be to add a script which does the equivalent to `pylint $(find test -type f -name '*.py')` and call that inside `tox.ini` instead. I also wonder if I should open a bug/suggestion against pylint for this.

nicoddemus committed:
Fixed assertions and other issues as reported by @The-Compiler

nicoddemus committed:
Renamed test to tests as suggested by @The-Compiler

nicoddemus commented:
About pylint wanting a package, perhaps you can at least inquire the reasoning behind this?

nicoddemus committed:
Merge remote-tracking branch 'origin/master' into layout

nicoddemus committed:
Removed xfail from test since issue has been fixed on master

The-Compiler committed:
Make lints run with adjusted test folder location.

For pylint we need a custom script; see
https://bitbucket.org/logilab/pylint/issue/512/

The-Compiler commented:
Sorry for the spam yesterday! I can confirm it looks fine now :+1:

I've opened [an issue for pylint](https://bitbucket.org/logilab/pylint/issue/512/) and added a small script in d3a92d505cb94c6df4db6139d1cb943ab7caf68c so all checkers run again.

I'll now fix the style issues so the checkers are quiet, take another quick look at the diff, and merge. Thanks for the big effort you're putting into this!

nicoddemus commented:
My pleasure! 

I gotta say, your test suite is very well written and complete. Congratulations! :smile:

I will wait for a new commit from you with the style issues fixed and then merge! 

Cheers, :beers:

The-Compiler committed:
Fix lint.

The-Compiler commented:
There we go! One more question before merging: It seems this also includes the GUI tests from #9 - is that intended? Should we remove them from here, and then merge #9 separately and move them while merging?

nicoddemus commented:
It wasn't intended, but I don't see any problem just merging this PR and close #9, as it was reviewed already.

nicoddemus commented:
Once @hackebrot also finishes reviewing, we can merge this in. :smile:

hackebrot commented:
Hey there,

apologies for my late reply. I've been on the road all day long.

I can review these changes tomorrow at the earliest given that the diff is massive. :worried: 

I understand if you do not want to wait for my review.

nicoddemus commented:
> apologies for my late reply. I've been on the road all day long.

No need to apologize, it's Sunday after all. :smile:

> I can review these changes tomorrow at the earliest given that the diff is massive. :worried:
> I understand if you do not want to wait for my review.

No worries, there's no rush.

Cheers, :beers:

hackebrot commented on MANIFEST.in:
```diff
@@ -6,7 +6,7 @@ graft icons
 graft doc/img
 graft misc
-include qutebrowser/test/testfile
+include qutebrowser/utils/testfile
```
To be perfectly honest I am not a big fan of having static test files in production code.

Shouldn't we rather have a folder inside `tests` for static test files?

The-Compiler replied:
The corresponding test is testing [qutebrowser.utils.utils.read_file](https://github.com/The-Compiler/qutebrowser/blob/master/qutebrowser/utils/utils.py#L68-L92), i.e. a function which loads a resource file from the qutebrowser package (or a subpackage) - so that'd kinda defeat the purpose in this case :wink:

hackebrot replied:
I see...fair enough :blush:

hackebrot commented on tests/browser/http/test_content_disposition.py:
```diff
@@ -20,11 +20,11 @@
 import os
-import unittest
 import logging
 
```
This is nitpicking at it's best...here's a redundant blank line.

The-Compiler replied:
Nitpicking is always appreciated! :smile:

Imports are grouped (according to [PEP8](https://www.python.org/dev/peps/pep-0008/#imports)) into stdlib/3rdparty/local imports, so that's fine.

hackebrot replied:
Haha :smile: 

In the **split** view there is a blank line between `import logging` and `import pytest`.

hackebrot commented on tests/browser/http/test_content_disposition.py:
```diff
@@ -140,53 +144,55 @@ def test_attonly(self):
         """'ATTACHMENT' only
 
         UA should offer to download the resource.
        """
```
Broken indentation for quotation marks.

hackebrot commented on tests/browser/http/test_content_disposition.py:
```diff
@@ -122,17 +126,17 @@ def test_inlwithasciifilenamepdf(self):
 
-class AttachmentTests(AttachmentTestCase):
 
+class TestAttachment:
```
Please inherit from `object`.

The-Compiler replied:
-1 - note qutebrowser is python3 only where inheriting from `object` isn't needed, and throughout the codebase also isn't done - so we should keep it that way for consistency :smile:

hackebrot replied:
There are some classes though that **do** inherit from `object`. :cry:

nicoddemus replied:
Yep, in fact when I started converting a few classes I subclassed them from `object` out of habit, then I realized we are using py3k and don't need that anymore (in fact I think it is no longer considered good style). :smile:

hackebrot commented on tests/browser/http/test_content_disposition.py:
```diff
@@ -140,53 +144,55 @@ def test_attonly(self):
     # we can't test attonly403 here.
 
-    def test_attonlyucase(self):
+    def test_attonlyucase(self, header_checker):
```
We should create an additional PR that renames these tests as they are pretty difficult to understand for someone who is new to the project. :worried:

The-Compiler replied:
Hmm. I see the reasoning, but the actual test cases are from [this page](http://greenbytes.de/tech/tc2231/), and like this (and explaining what the test does in the docstring) it's easy to see which test corresponds to which one on this page.

An alternative would of course be to give them better names and mention the original name in the docstring.

hackebrot replied:
Ah okay. Thank you for the reference! :bow:

hackebrot commented on tests/browser/http/test_content_disposition.py:
```diff
@@ -345,10 +355,10 @@ def test_attwithfnrawpctencaq(self):
-        self._check_filename(r'attachment; filename="foo-%\41.html"',
-                             'foo-%41.html')
+        header_checker.check_filename(r'attachment; filename="foo-%\41.html"',
+                                      'foo-%41.html')
```
Unwanted whitespaces :question:

The-Compiler replied:
Looks good to me (hanging indentation according to [PEP8](https://www.python.org/dev/peps/pep-0008/#indentation))

hackebrot replied:
My bad. :no_mouth:

It's the longer method invocation that causes an additional indentation level.

hackebrot commented on tests/browser/http/test_content_disposition.py:
```diff
@@ -585,11 +598,11 @@ def test_attabspathwin_unix(self):
-        self._check_filename(r'attachment; filename="\\foo.html"',
-                             r'\foo.html')
+        header_checker.check_filename(r'attachment; filename="\\foo.html"',
+                                      r'\foo.html')
```
Whitepaces :smile_cat:

hackebrot commented on tests/browser/http/test_content_disposition.py:
```diff
@@ -844,12 +871,12 @@ def test_attnewandfn(self):
 
-class RFC2047EncodingTests(AttachmentTestCase):
 
+class TestRFC2047Encoding:
```
Please inherit from `object`.

hackebrot commented on tests/browser/http/test_http.py:
```diff
@@ -23,46 +23,39 @@
 
 
-class ParseContentTypeTests(unittest.TestCase):
+class TestParseContentType:
```
Again I'm voting for new style classes. :smirk:

hackebrot commented on tests/browser/test_tabhistory.py:
```diff
@@ -0,0 +1,139 @@
+from qutebrowser.utils import qtutils
+
+
+class TestSerializeHistory:
```
New style classes :smile:

hackebrot commented on tests/browser/test_tabhistory.py:
```diff
@@ -0,0 +1,139 @@
+    ]
+
+    @pytest.fixture(autouse=True)
+    def setup(self, webpage):
```
Shouldn't we rather use `setup_method` without the decorator :question: 

See https://pytest.org/latest/xunit_setup.html#method-and-function-level-setup-teardown

nicoddemus replied:
Problem with `setup_method` is that they can't receive fixtures as arguments, which is needed in several of the `setup` methods of classes I converted (in this case for example, it uses the `webpage` fixture). 

Those few that don't need any fixtures could instead declare `setup_method(self, method)` I suppose, but I rather use fixtures in all of them for consistency and because I believe fixtures are superior. :smile:

hackebrot commented on tests/browser/test_tabhistory.py:
```diff
@@ -0,0 +1,139 @@
+
+    def test_userdata(self):
+        """Check if all user data has been restored to self.user_data."""
+        for item, user_data in zip(self.ITEMS, self.user_data):
```
What do you think of the following?

``` python
assert self.ITEMS == self.user_data
```

I think that `py.test` will still be able to create a comprehensible error message.

nicoddemus replied:
Oh, totally agree, didn't realize that at the time of the conversion. :smile:

Had to do it slightly different though:

``` python
    def test_userdata(self):
        """Check if all user data has been restored to self.user_data."""
        userdata_items = [item.user_data for item in self.ITEMS]
        assert userdata_items == self.user_data
```

hackebrot commented on tests/browser/test_tabhistory.py:
```diff
@@ -0,0 +1,139 @@
+        assert self.history.itemAt(i).title() == item.title
+
+
+class TestSerializeHistorySpecial:
```
New style classes.

The-Compiler commented on tox.ini:
```diff
@@ -31,6 +33,9 @@ deps =
     pytest-cov==1.8.1
+    pytest-capturelog==0.7
+    pytest-qt==1.3.0
+    pytest-mock==0.4.2
```
Those are all already included by the `{[testenv:unittests]deps}` above, so shouldn't be added here again or pip will complain about multiple requirements.

nicoddemus replied:
Oops, you're right! Thanks. :smile:

hackebrot commented on tests/browser/test_webelem.py:
```diff
@@ -87,17 +86,17 @@ def _style_property(name, strategy):
 
 
-class WebElementWrapperTests(unittest.TestCase):
+class TestWebElementWrapper:
```
I'm skipping the **new-style-class** comments for the rest of the diff :exclamation:

hackebrot commented on tests/browser/test_webelem.py:
```diff
@@ -360,189 +360,191 @@ def test_multiple_levels(self):
+    def stub_config(self, stubs):
+        """Fixture to create a config stub with an input section."""
+        config = stubs.ConfigStub({'input': {}})
+        with mock.patch('qutebrowser.browser.webelem.config', new=config):
```
Not too sure about the usage of `mock`. Shouldn't we rather try to remove all usages of `unittest` and use `monkeypatch` or `pytest-mock`[1]. By the looks of it the latter is just a helper but I feel we should use the same concepts throughout the testsuite.

Very much looking for your feedback! :relaxed: 

[1] https://pypi.python.org/pypi/pytest-mock

nicoddemus replied:
That must have been a brain fart from my part. Fixed! :smile:

Completely agree that for tests converted to pytest style we should use `monkeypatch` (for simple substitutions) or `pytest-mock` (when you need the power of `Mock` objects).

hackebrot commented on tests/config/test_config.py:
```diff
@@ -116,50 +112,50 @@ def test_interpolation_cross_section(self):
 
 
-class DefaultConfigTests(unittest.TestCase):
+class TestDefaultConfig:
```
Are we missing an assert here or is this supposed to be a **smoke-test**?

nicoddemus replied:
I think so. @The-Compiler?

The-Compiler replied:
Yeah, it's basically a smoke test - it verifies if the built-in default config validates (i.e. doesn't raise `qutebrowser.config.configexc.ValidationError`.

Once I made a typo in a default config value, didn't notice it because I of course already had a config, and all new users got an exception right away :blush:

hackebrot commented on tests/config/test_config.py:
```diff
@@ -169,40 +165,32 @@ def test_default_config(self):
-            'XDG_CONFIG_HOME': self.conf_path,
-            'XDG_DATA_HOME': self.data_path,
-            'XDG_CACHE_HOME': self.cache_path,
+            'XDG_CONFIG_HOME': str(self.conf_path),
```
I'm wondering if we are running into problems with the `str` cast as Python 3 uses unicodes by default :worried:

nicoddemus replied:
Shouldn't be a problem, since Python 3 paths are unicode. :wink:

hackebrot commented on tests/config/test_configtypes.py:
```diff
@@ -70,72 +69,80 @@ def __repr__(self):
-        self.assertIn('foo', vv)
-        self.assertNotIn("baz", vv)
+        assert 'foo' in vv
+        assert "baz" not in vv
```
Mixed usage of single and double quotes. Feel free to ignore this for now as it was like this before.

nicoddemus replied:
No problem, changed.

hackebrot commented on tests/config/test_configtypes.py:
```diff
@@ -328,38 +333,39 @@ class BoolTests(unittest.TestCase):
         self.t = configtypes.Bool()
 
-    def test_transform(self):
+    _ALL_FORMS = []
```
Should we have a method for this and call it in `pytest.mark.parametrize` :question:

nicoddemus replied:
Changed it to an inline list comprehension, like the other cases in this module.

hackebrot commented on tests/config/test_configtypes.py:
```diff
@@ -861,30 +870,31 @@ def test_validate_both_range_perc(self):
 
-    def setUp(self):
+    @pytest.fixture(autouse=True)
+    def setup(self, mocker, stubs):
```
Correct me if I'm wrong but here we are using `pytest-mock` right?

nicoddemus replied:
Exactly. :smile:

hackebrot commented on tests/config/test_configtypes.py:
```diff
@@ -1124,15 +1151,16 @@ class FontTests(unittest.TestCase):
     INVALID = ['green "Foobar Neue"', 'italic green "Foobar Neue"',
                'bold bold "Foobar Neue"', 'bold italic "Foobar Neue"'
-               'bold', '10pt 20px "Foobar Neue"']
+                                          'bold', '10pt 20px "Foobar Neue"']
```
Broken indentation

nicoddemus replied:
I think that indentation broke because the `'bold'` at the bottom is actually part of the `'bold italic "Foobar Neue"'` string in the line above it, since there is no colon between them which means Python joins both strings. Is this intended @The-Compiler? Either way, changing the string above to `'bold italic "Foobar Neue" bold'` or leaving out the trailing `bold` the test passes.

The-Compiler replied:
Oh, weird! I don't think that `bold` is supposed to be there, as the aim of the test is to test the `bold italic ...`. I'll push a commit which fixes it and uses one line per test because the quotes quickly get confusing.

hackebrot commented on tests/config/test_configtypes.py:
```diff
@@ -2087,41 +2116,43 @@ def test_validate_empty_none_ok(self):
     def test_transform_empty(self):
         """Test transform with an empty value."""
-        self.assertEqual(self.t.transform(''), None)
+        assert self.t.transform('') is None
```
I may be wrong but this is not exactly the same, is it?

`self.assertEqual` is different from `self.assertIs`

nicoddemus replied:
You're correct, but it is recommended to compare against `None` using the `is` operator. In fact I believe this was supposed to be `self.assertIsNone(self.t.transform(''))`, so I went ahead and changed it. :smile:

The-Compiler replied:
For `None`, `is` and `==` happens to do the same, but of course `==` is discouraged. No idea why I wrote it like this :laughing:

hackebrot commented on tests/utils/test_utils.py:
```diff
@@ -244,122 +242,117 @@ class FormatSizeTests(unittest.TestCase):
-    def test_format_size(self):
+    KILO_TESTS = [(999, '999.00'), (1000, '1.00k'), (1010, '1.01k')]
+
+    @pytest.mark.parametrize('size, out', TESTS)
```
:+1:

hackebrot commented:
Great work @nicoddemus @The-Compiler :exclamation: 

I feel that we should straighten out the usage of `mocker` (fixture of `pytest-mock`) and `mock` of Python 3 `unittest`.

Other than that I am voting for replacing 

``` python
@pytest.fixture(autouse=True)
def setup(self):
    pass
```

with

``` python
def setup_method(self):
    pass
```

Once again, well done friends! :smiley: :+1:

nicoddemus committed:
Remove "object" subclassing from Test classes missed initially

nicoddemus committed:
Fix small docstring issues

nicoddemus committed:
Compare full lists instead of looping over items as suggested by @hackebrot

nicoddemus committed:
Remove spurious dependencies from tox.ini as pointed out by @The-Compiler

nicoddemus committed:
Use mocker fixture instead of unittest.mock

As pointed out by @hackebrot

nicoddemus committed:
Use single-quoted string for consistency

nicoddemus committed:
Use inline list comprehension for parametrize expression

nicoddemus commented:
Thanks for the thorough review @hackebrot! 

To summarize your last points, I fixed the `mock` usage in favor of `mocker` as you requested, and about `setup_method` please see [this comment](https://github.com/hackebrot/qutebrowser/pull/16#discussion_r27940482).

Cheers, :beers:

The-Compiler committed:
Fix string concatenation and indenting for INVALID.

The-Compiler committed:
Add some more invalid testcases to TestFont.

The-Compiler commented:
I fixed the weird string concatenation for the invalid fonts and [added some more test cases](https://github.com/hackebrot/qutebrowser/commit/679ffa452a4158b69b6d8c25aceec72472fb0d12) there inspired by that.

The buildbot also seems happy and green, so I think I can merge this, correct?

Thanks for the great work and thorough review - it's really a joy working with you guys!

The-Compiler committed:
Another workaround for a pylint bug.

nicoddemus commented:
> The buildbot also seems happy and green, so I think I can merge this, correct?

:+1: from me.

> Thanks for the great work and thorough review - it's really a joy working with you guys!

Likewise! :smile:

hackebrot commented:
:+1:
